### PR TITLE
For the INTT hit carry-over issue in streaming, deepen the m_InttRawHitMap size, and discarding non-collision crossings

### PR DESCRIPTION
**Pull request by ChengWeiShih**
## Types of changes
[comment]: <> ( What types of changes does your code introduce? Put an `x` in all the boxes that apply: )
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work for users)
- [ ] Requiring change in macros repository (Please provide links to the macros pull request in the last section)
- [x] I am a member of [GitHub organization of sPHENIX Collaboration](https://github.com/orgs/sPHENIX-Collaboration/people), EIC, or ECCE (contact Chris Pinkenburg to join)

## Summary
1. The PR recovers missing hits.
2. The PR represents the ultimate way to minimize the INTT carry-over hit issue in streaming readout (Discussion in https://indico.bnl.gov/event/33369/#5-cheng-weis-update)
3. The PR has a function to remove hits from non-collision crossings. Those should be excluded from the physics analysis.

### For recovering missing hits
- Deepen the size of `std::map<uint64_t, InttRawHitInfo> m_InttRawHitMap;`
  - It is found that hits are missing because the m_InttRawHitMap in the `Fun4AllStreamingInputManager` is not deep enough. 
  - The size of `m_InttRawHitMap` in `Fun4AllStreamingInputManager` is controlled by `GetSomeMoreEvents()` in `SingleInttPoolInput`.
  - The requirement is now `if (bcliter.second <= localbclk + 120 * 40)`, ensuring > 40-strobe depth in the map. 
  - The memory usage is below 2 GB for one .evt file.
  
----
### For the INTT carry-over hit issue in streaming
- Public function added in `Fun4AllStreamingInputManager`:
  - `EnableInttHitDuplication(bool)`:  Enable/disable the mitigation (defaults to disabled)
  - `InttHitCarryOverShiftMaxMultiple(int)`: Set the strobe search depth  (defaults to four future strobes)
  - `SetIsRejectInttNoiseCrossings(bool)`: Enable/disable removing INTT hits from non-collision crossings (defaults to disable)
  - `SetIsInttStreaming(bool)`: Set if INTT is in the streaming readout mode (defaults to true)
  - `SetIsDuplicateInttFPHXBCOResetHit(bool)`: enable/disable duplicating INTT hits with FPHXBCO reset to the values defined by `std::vector<int> m_InttResetFphxBcoVec` (defaults to true)
  - `SetInttResetFphxBcoVec(std::vector<int>)`: Set the `std::vector<int> m_InttResetFphxBcoVec`(default {0 to 5})
  - `SetInttStreamingSignalCrossing(std::pair<int,int>)`: Set the INTT FPHXBCO range with collision, `std::pair<int,int> m_InttStreamingSignalCrossing`. (defaults to {6, 116}). 

- Implementation in the `Fun4AllStreamingInputManager` for the INTT carry-over hit issue in the streaming readout
  - Inclusive correction following the half-ladder ordering 
    - A map, `std::map<std::string, int> m_InttRawHitCount_FEE;`, is prepared for every F4A event to count the number of hits for each half-ladder in each crossing of the five strobes in a F4A event (1 GL1BCO) 
      - The key of the map:  `BCOFULL_FPHXBCO_FELIXID_FELIXChannel`.
    - For entries in the map with FPHXBCO ranging from 6 to 116, we look for hits matched to the following text, duplicate those hits, and reset their BCOfull. For hits to be duplicated with FPHXBCO 0 to 5, we reassign their FPHXBCO. 
      - BCOfull+k*120_FELIXID_FEE_FPHXBCO
      - BCOfull+k*120_FELIXID_FEE_AAA (AAA: 0 to 5)
    - k <= N, where N is the searching depth, the `m_InttHitCarryOverShiftMaxMultiple`.
    - The demonstration of "Inclusive correction" can be found in https://indico.bnl.gov/event/33369/#5-cheng-weis-update
  - Optional: boolean variable `m_IsDuplicateInttFPHXBCOResetHit` in the `Fun4AllStreamingInputManager` to enable/disable duplicating hits with FPHXBCO reset to 0 to 5.
  - The implementation also incorporates the correction method for the triggered case to be implemented in the future.
  
----
### For discarding hits from non-collision crossings.
- Prevent hits from no-collision crossings from being saved to the INTTRawHit container
  - Only for the streaming readout (m_IsInttStreaming ==true)
  - INTT strobe is synchronized with the RHIC circulation. The RHIC bunch crossings should only happen in FPHXBCO 6 to 116. Therefore, a hit with an FPHXBCO outside the range should not be a signal hit.
  - A boolean variable `m_IsRejectInttNoiseCrossings` in `Fun4AllStreamingInputManager` to remove those hits after the carry-over hit correction.

## TODOs (if applicable)

The current implementation for hit duplication works only with the INTT streaming data. In the near future, we will implement the carry-over hit correction for the INTT-triggered data, which will be incorporated into the current implementation.

## Links to other PRs in macros and calibration repositories (if applicable)

Motivation / context:  
This PR addresses an INTT streaming readout carry-over issue where hits can be missed across strobes. It extends the effective recovery depth by duplicating INTT raw hits across future BCOs (within a configurable carry-over window) when half-ladder/FPHX association is available. AI-generated summaries can make mistakes—please use best judgment when reading.

Key changes:
- Added new public configuration/setter APIs in `Fun4AllStreamingInputManager` to control INTT streaming behavior, including:
  - enabling/disabling INTT hit duplication (`EnableInttHitDuplication`)
  - carry-over search depth (`InttHitCarryOverShiftMaxMultiple`, with validation that it must be non-negative)
  - optional rejection of “noise crossings” during streaming save/duplication (`SetIsRejectInttNoiseCrossings`, `SetIsInttStreaming`) using a configurable signal-crossing FPHX_BCO range (`SetInttStreamingSignalCrossing`)
  - duplication rules for reset hits and their allowed FPHX_BCO values (`SetIsDuplicateInttFPHXBCOResetHit`, `SetInttResetFphxBcoVec`)
- Updated `Fun4AllStreamingInputManager::FillIntt` to implement carry-over duplication:
  - when duplication is enabled, rebuilds a per-event/per-FEE hit-count map (`m_InttRawHitCount_FEE`) keyed by `(bco, FPHX_BCO, server/packetid, fee/channel)`, limited to the active BCO selection window
  - first pass: saves original hits into `INTTRAWHIT`, optionally skipping hits whose `FPHX_BCO` lies outside the configured signal-crossing range (when noise rejection is enabled and INTT streaming is active)
  - duplication pass: for each half-ladder key, searches forward in `m_InttRawHitMap` for future strobes at `future_bco = base_bco + m_InttHitCarryOverShift * search_i` (with `search_i = 1..m_InttHitCarryOverShiftMaxMultiple`), and copies matching hits from the same server/fee/channel when:
    - `FPHX_BCO` matches the half-ladder’s `FPHX_BCO`, or
    - reset duplication is enabled and `FPHX_BCO` is contained in `m_InttResetFphxBcoVec`
  - copied hits have their `bco` and `FPHX_BCO` overwritten to the target half-ladder values; noise-crossing rejection is also applied to the half-ladder keys during this duplication pass
- Added/expanded INTT diagnostics in `FillIntt`:
  - one-time state dump (guarded by `m_Intt_print_count`)
  - additional verbose printing (e.g., after GL1 BCO matching and for the per-key hit-count map) under `Verbosity()` thresholds
- Relaxed an INTT “stuck beam clock” erasure condition in `SingleInttPoolInput` by widening the acceptable BCL gap (compare against `localbclk + 120 * 40` instead of `localbclk`).

Potential risk areas:
- Reconstruction/output behavior changes to the contents of `INTTRAWHIT` due to added hit duplication and optional exclusion of out-of-range (non-collision) crossings.
- Sensitivity to configuration correctness (signal-crossing range, shift multiple, and reset-FPHXBCO vector), which directly impacts recovered/duplicated hit counts.
- Performance impact from additional map building and forward searching across strobes (tunable via the new carry-over/max-multiple configuration).
- Thread-safety considerations: new mutable configuration/state in `Fun4AllStreamingInputManager` could be risky if the manager is accessed concurrently without external synchronization.

Possible future improvements:
- Extend and validate the same carry-over/duplication strategy for triggered (non-streaming) data as planned.
- Add targeted tests/QA for edge cases: reset-FPHXBCO duplication behavior, boundary values for the signal-crossing range, and carry-over windows where future strobes are missing.
- Consider refactoring the duplication key parsing/counting logic to reduce complexity and improve maintainability.

### CODE REVIEW

**Comment by hrjheng** [comment]
I closed my PR https://github.com/sPHENIX-Collaboration/coresoftware/pull/4290 as discussed in the dNdEta analysis meeting. Below are a few notes for this PR

Cheng-Wei can provide the full explanation, but here is my understanding of the differences between the two implementations:
1. Both implementations use the same two-pass duplication strategy. The first duplication pass is identical (at least conceptually).
2. Cheng-Wei pointed out that, although my second duplication pass correctly reassigns the FPHX BCO, it does not update `BCO_full`. As a result, the duplicated hits are not actually assigned to the intended strobe. Cheng-Wei's implementation corrects this by updating `BCO_full` together with the FPHX BCO.
3. Cheng-Wei's implementation also duplicates hits with FPHX BCO values from 0 through 5, rather than only FPHX BCO = 0. The hypothesis is that when the FPHX BCO is reset, it may be reset to any value between 0 and 5, randomly or not. This hypothesis has not yet been verified.

To test 3, Cheng-Wei will compare the INTT doublet distributions as a function of FPHX BCO using two configurations:
* duplicating only hits with FPHX BCO = 0 (with the corrected `BCO_full` assignment), and
* duplicating hits with FPHX BCO values from 0 to 5.
If the two approaches produce significantly different results, that would provide evidence supporting the hypothesis that the reset FPHX BCO can take values other than 0. If there is little or no difference, the additional duplication may not be necessary.

**Comment by ChengWeiShih** [comment]
- Regarding @hrjheng 's first point. The idea of the first pass is the same, but the number of hits being duplicated is different.
  - Hao-Ren's PR: Duplicating every hit from the future strobes. (taking the whole FELIX server, 14 half-ladders * 120 FPHXBCOs, as a unit)
  Cheng-Wei's PR: Take a `half-ladder in a crossing` as a unit. Which means the number of INTT hits for each half-ladder at each crossing is counted before correction. And duplicate hits only for the half-ladders in a given crossing of the strobe that have a hit.
  - Difference: While Hao-Ren's approach duplicates more hits, a crossing with entire fake hits can potentially be artificially created. CW's method has an assumption that at least one hit in a half-ladder in a crossing is correctly associated; fake crossings therefore are minimized.
  
- For the second point, the second pass, I think the missing pieces in Hao-Ren's PR are:  
  1. Missing FPHXBCO reassignment, which should be done by `copied_hit->set_FPHX_BCO(The_correct_FPHXBCO);`
  2. Incorrect `bco_full` search and reassignment, as `const uint64_t target_refbco = bco % m_InttHitCarryOverShift;` depends on when the data taking started. I write down an example:
      1.  `uint64_t bco = 608318921053;`
      2. `target_refbco = 608318921053 % 120 = 13` 
      3.  `shift = 120 * N`. Assume N = 2, `shift = 240`.
      4. `reset_shift = shift - target_refbco = 240 - 13 = 227`.
      5.  `reset_source_bco = bco + reset_shift = 608318921053 + 227 = 608318921280`.
      - There is no strobe with `bco = 608318921280`, as the `bco` of the next strobe should be `608318921053 + 120 = 608318921173` and then `608318921293`, and so on.
      - The second pass in the PR would find nothing.

- For the third point, I will do the study.

- I would like to point out that @hrjheng should also have credit. I don't mean to take his credit. His first look, trial, and first implementation are inspiring and can largely duplicate hits. I learned from him that the implementation can be done in the `StreamingManager`. My implementation is inspired by his PR.

File: offline/framework/fun4allraw/SingleInttPoolInput.cc [hunk 1]
```diff
@@ -466,7 +466,7 @@ bool SingleInttPoolInput::GetSomeMoreEvents(const uint64_t ibclk)
   std::set<int> toerase;
   for (auto bcliter : m_FEEBclkMap)
   {
-    if (bcliter.second <= localbclk)
+    if (bcliter.second <= localbclk + 120 * 40)
```
[the remark this answers is not in the document]
Comment (ChengWeiShih) [reply offline/framework/fun4allraw/SingleInttPoolInput.cc:469]: No, this wouldn't work, as this is related to how the PacketVector arranges the data.
